### PR TITLE
Support multiple account verification methods

### PR DESCRIPTION
**Pull request by robertabcd**
**A new `verifydb` table served by `regmaild` that stores unique verification key.**

The current `emaildb` table stores `userid` and `email` only. Since `userid` may be reused after it expires, the server determines whether the entry is up-to-date by comparing `email` with the one in passwd file.

As `userec_t` is full, we can't add more fields to support other verification methods. (i.e. no comparing with passwd file) The new `verifydb` table uses `(userid, generation)` to identify each user. `generation` is the `firstlogin` field in `userec_t`.

To support multiple verification at the same, PK is set to `(userid, generation, vmethod)`.

Email is now dual-write to verifydb.

### CODE REVIEW

File: daemon/regmaild/regmaild.c [hunk 1]
```diff
@@ -100,52 +147,47 @@ int main(int argc, char *argv[])
...
+            fprintf(stderr, "\ncannot execute statement: %s\n",
+                    sqlite3_errmsg(Db));
             break;
         }
         sqlite3_reset(Stmt);
 
-        valids ++;
+        valids++;
         if (valids % 10 == 0)
-            fprintf(stderr, "%d/%d (valid: %d)\r",
```
Comment (wens) [inline daemon/regmaild/regmaild.c]: Could you split out the error number fixes and style fixes into separate patches when possible?
That would make it easier to read.

File: daemon/regmaild/regmaild.c [hunk 2]
```diff
@@ -73,6 +73,52 @@ regmaildb_open(sqlite3 **Db, const char *fpath) {
...
+}
+
+static int
+emaildb_filter_user(const userec_t *u)
+{
+    if (strlen(u->userid) < 2 || strlen(u->userid) > IDLEN)
+        return 0;
+    if (strlen(u->email) < 5)
+        return 0;
+    return 1;
```
Comment (wens) [inline daemon/regmaild/regmaild.c]: If all you are returning is success and failure, maybe use `bool` as return type explicitly?

File: daemon/regmaild/regmaild.c [hunk 3]
```diff
@@ -73,6 +73,52 @@ regmaildb_open(sqlite3 **Db, const char *fpath) {
...
+}
+
+static int
+emaildb_bind_args(sqlite3 *db, sqlite3_stmt *stmt, const userec_t *u)
+{
+    if (sqlite3_bind_text(stmt, 1, u->userid, -1, SQLITE_STATIC) != SQLITE_OK ||
+        sqlite3_bind_text(stmt, 2, u->email, -1, SQLITE_STATIC) != SQLITE_OK)
+    {
+        fprintf(stderr, "\nError binding args: %s\n", sqlite3_errmsg(db));
```
Comment (wens) [inline daemon/regmaild/regmaild.c]: Get rid of the leading newline?

File: daemon/regmaild/regmaild.c [hunk 4]
```diff
@@ -107,6 +109,44 @@ emaildb_filter_user(const userec_t *u)
...
+    // vkey will be lower-cased in sql stmt.
+    char vkey[EMAILSZ+16];
+    snprintf(vkey, sizeof(vkey), "email:%s", u->email);
+
+    if (sqlite3_bind_text(stmt, 1, u->userid, -1, SQLITE_STATIC) != SQLITE_OK ||
+        sqlite3_bind_int64(stmt, 2, u->firstlogin) != SQLITE_OK ||
+        sqlite3_bind_int64(stmt, 3, 0) != SQLITE_OK ||
+        sqlite3_bind_text(stmt, 4, vkey, -1, SQLITE_TRANSIENT) != SQLITE_OK)
+    {
+        fprintf(stderr, "\nError binding args: %s\n", sqlite3_errmsg(db));
```
Comment (wens) [inline daemon/regmaild/regmaild.c]: Drop the leading newline.

**Comment by wens** [comment]
So this supports multiple options for verification, but not multiple concurrent options, i.e. both email and mobile are verified. Am I right?

**Comment by robertabcd** [comment]
The schema now supports multiple verification methods for an user at the same time. Also dual-write implemented.

Other verification methods will need more refactoring, and will come in separate PRs.

File: daemon/regmaild/regmaild.c [hunk 5]
```diff
@@ -278,12 +272,206 @@ regmaildb_update_email(const char * userid, int userid_len, const char * email,
...
+            fprintf(stderr, "verifydb count: skipped old record: %s "
+                    "(generation: db %ld, current %d)\n",
+                    u.userid, generation, u.firstlogin);
+            continue;
+        }
+
+        if (strcasecmp(userid, req->userid) == 0)
+            rep->count_self++;
+        else
+            rep->count_other++;
```
Comment (wens) [inline daemon/regmaild/regmaild.c:429]: This is almost certainly not going to match, because of the generation check above.
What you might want is to place the generation check inside the userid match case.

File: daemon/regmaild/regmaild.c [hunk 6]
```diff
@@ -278,12 +272,206 @@ regmaildb_update_email(const char * userid, int userid_len, const char * email,
...
+    }
+
+    rep->status = VERIFYDB_OK;
+
+end:
+    if (r != SQLITE_OK)
```
Comment (wens) [inline daemon/regmaild/regmaild.c:435]: You can use multiple goto targets, one for each failure case.
That way you can get rid of some the if tests below.
This is used extensively in the kernel.

File: daemon/regmaild/regmaild.c [hunk 7]
```diff
@@ -278,12 +272,206 @@ regmaildb_update_email(const char * userid, int userid_len, const char * email,
...
+        fprintf(stderr, "sqlite3: %d %s\n", r, sqlite3_errmsg(Db));
```
Comment (wens) [inline daemon/regmaild/regmaild.c:436]: Are the standard sqlite3 error messages specific enough that you can tell which function above caused it?
If not, it might be best to have more detailed error messages directly following the error.

File: daemon/regmaild/regmaild.c [hunk 8]
```diff
@@ -516,87 +704,204 @@ regcheck_ambiguous_id(const char *userid)
 //
 // operation: set,count
 
-static void
-err_request(regmaildb_req *req, int fd, struct event *ev)
+static bool
+check_null_terminated(const char *s, size_t sz)
```
Comment (wens) [inline daemon/regmaild/regmaild.c:812]: `return !!memchr(s, '\0', sz);` ?

File: daemon/regmaild/regmaild.c [hunk 9]
```diff
@@ -516,87 +704,204 @@ regcheck_ambiguous_id(const char *userid)
...
+        case REGCHECK_REQ_AMBIGUOUS:
+            // No checks.
+            break;
+
+        default:
+            fprintf(stderr, "unknown request: op=[%d]\n", req->operation);
+            return false;
+    }
+    return true;
+}
```
Comment (wens) [inline daemon/regmaild/regmaild.c:847]: This change and other cleanup of regmaildb code could be pulled out into a separate "preparatory" commit.
In a way you will prepare existing code and transform it into a structure that can accept adding the new verifydb
stuff with minimal changes.

File: daemon/regmaild/regmaild.c [hunk 10]
```diff
@@ -516,87 +704,204 @@ regcheck_ambiguous_id(const char *userid)
...
+    int8_t *rest = (int8_t *)&storage;
+    rest += sizeof(*header);
 
     assert(ev);
 
-    if ( (event & EV_TIMEOUT) ||
-        !(event & EV_READ) ||
-         toread(fd, &req, sizeof(req)) != sizeof(req) ||
-         req.cb != sizeof(req))
+    if ((event & EV_TIMEOUT) ||
```
Comment (wens) [inline daemon/regmaild/regmaild.c]: Personally I like the style of "matching the opening parenthesis", as it allows one to quickly see which parenthesis level it is part of.

File: daemon/regmaild/regmaild.c [hunk 11]
```diff
@@ -516,87 +704,204 @@ regcheck_ambiguous_id(const char *userid)
...
+            toread(fd, header, sizeof(*header)) != sizeof(*header) ||
+            header->cb < sizeof(*header) ||
+            header->cb > sizeof(storage) ||
+            toread(fd, rest, header->cb - sizeof(*header)) != header->cb - sizeof(*header))
     {
-        fprintf(stderr, "error: corrupted request.\r\n");
-        close(fd);
-        free(ev);
-        return;
+        fprintf(stderr, "error: corrupted request.\n");
```
Comment (wens) [inline daemon/regmaild/regmaild.c]: Either leave the carriage return alone for now, or make a patch removing it from all log messages.
It is unrelated to what the commit log says the change is supposed to do.
(Why were there carriage returns in the first place...)

File: include/daemons.h [hunk 12]
```diff
@@ -31,6 +56,60 @@ typedef struct
...
 }   regmaildb_req;
 
+#define VERIFYDB_VKEY_SIZE (160)
+
+// verifydb status codes.
+#define VERIFYDB_OK (0)
+#define VERIFYDB_ERROR (-1)
+
+// verifydb vmethod keys.
+#define VMETHOD_UNSET (0)
```
Comment (wens) [inline include/daemons.h]: Use an `enum`.

File: include/daemons.h [hunk 13]
```diff
@@ -31,6 +56,60 @@ typedef struct
...
+
+typedef struct {
+    size_t cb;
```
Comment (wens) [inline include/daemons.h:74]: This deserves a better name. Based on your usage, the shortest unambiguous name I can think of is `bytecount`.

**Review by wens — COMMENTED** [review]
First commit reviewed.
The commit message should be clear that it is introducing a framework, and does not actually support any vmethods at this point.

File: daemon/regmaild/regmaild.c (hunk 5 above)
Comment (robertabcd) [reply daemon/regmaild/regmaild.c:429]: This will match if another user is currently using the requested `(vmethod, vkey)`.

File: daemon/regmaild/regmaild.c (hunk 6 above)
Comment (robertabcd) [reply daemon/regmaild/regmaild.c:435]: I think this will become very hard to follow esp. in `verifydb_get`. As the performance isn't of concern here, `if` wouldn't hurt and I feel a single bail out point would be easier to follow.

File: daemon/regmaild/regmaild.c (hunk 7 above)
Comment (robertabcd) [reply daemon/regmaild/regmaild.c:436]: I won't say which function but it will contain some details in `errmsg` which is enough to know which function call failed.

File: daemon/regmaild/regmaild.c (hunk 8 above)
Comment (robertabcd) [reply daemon/regmaild/regmaild.c:812]: Done.

File: daemon/regmaild/regmaild.c (hunk 9 above)
Comment (robertabcd) [reply daemon/regmaild/regmaild.c:847]: Separated out.

File: daemon/regmaild/regmaild.c (hunk 10 above)
Comment (robertabcd) [reply daemon/regmaild/regmaild.c]: Done.

File: daemon/regmaild/regmaild.c (hunk 11 above)
Comment (robertabcd) [reply daemon/regmaild/regmaild.c]: Made a separate commit that removes "\r".

File: include/daemons.h (hunk 12 above)
Comment (robertabcd) [reply include/daemons.h]: Done.

File: include/daemons.h (hunk 13 above)
Comment (robertabcd) [reply include/daemons.h:74]: This file uses `cb` across the board. I think we should keep the name for consistency purpose, unless we want to change them all.

File: daemon/regmaild/regmaild.c [hunk 14]
```diff
@@ -526,34 +548,32 @@ client_cb(int fd, short event, void *arg)
...
     {
         case REGMAILDB_REQ_COUNT:
-            if (!*req.userid || !*req.email) {
-                err_request(&req, fd, ev);
-                return;
-            }
-            ret = regmaildb_check_email(req.email, strlen(req.email), req.userid);
+        {
+            if (!validate_regmaildb_request(&req))
+                goto end;
```
Comment (wens) [inline daemon/regmaild/regmaild.c]: I believe you can move this outside the switch / case.
After that you can either drop the default case, or mark it as not reached.

**Comment by wens** [comment]
"regmaild: Error and exit if DB cannot be opened. " looks like it should come before "regmaild: Use g_Db in handler." to preserve working functionality during bisection.

File: include/daemons.h [hunk 15]
```diff
@@ -31,6 +31,19 @@ typedef struct
...
+    int operation;
+} regmaildb_req_header;
+
+// Storage struct that fits all possible messages.
+typedef struct {
+    union {
+        regmaildb_req_header header;
+        regmaildb_req regmaildb;
+    } m;
+} regmaildb_req_storage;
```
Comment (wens) [inline include/daemons.h:117]: There might be a better way to do this:

```
typedef struct {
    size cb;
    int operation;
} regmaildb_req_header;

typedef struct {
    regmaildb_req_header; /* not sure if anonymous nested structs work this way */
    [request type-specific fields ... ]
} regmaildb_req;
```

But let's try that later.

File: daemon/regmaild/regmaild.c (hunk 14 above)
Comment (robertabcd) [reply daemon/regmaild/regmaild.c]: Um, in later commit I added `validate_verifydb_request` and I would have to move them back. I'd like to just keep them like this.

File: include/daemons.h (hunk 15 above)
Comment (robertabcd) [reply include/daemons.h:117]: This doesn't work. I notice I don't need to give a name to `union`, so I've removed it. This lets us do `req_storage.header` etc.

**Comment by robertabcd** [comment]
> "regmaild: Error and exit if DB cannot be opened. " looks like it should come before "regmaild: Use g_Db in handler." to preserve working functionality during bisection.

Right, reordered.

**Comment by robertabcd** [comment]
Accidentally merged two commits together. Reuploaded.

File: include/daemons.h (hunk 15 above)
Comment (wens) [reply include/daemons.h:117]: Just curious, why doesn't it work though?

File: daemon/regmaild/regmaild.c [hunk 16]
```diff
@@ -270,6 +276,202 @@ regmaildb_update_email(const char * userid, int userid_len, const char * email,
...
+
+    buf = evbuffer_new();
+    if (!buf)
+        goto end;
+    while (sqlite3_step(Stmt) == SQLITE_ROW) {
+        const int64_t timestamp = sqlite3_column_int64(Stmt, 0);
+        const int32_t vmethod = sqlite3_column_int64(Stmt, 1);
+        const char *vkey = (const char *)sqlite3_column_text(Stmt, 2);
+        if (!vkey)
+            continue;  // Bad record.
```
Comment (wens) [inline daemon/regmaild/regmaild.c]: Can't you just filter these out with the WHERE clause?
This way you could also get rid of `evbuffer_prepend()` and
instead queue up everything in order, or even drop `evbuffer` altogether.

File: daemon/regmaild/regmaild.c [hunk 17]
```diff
@@ -270,6 +276,202 @@ regmaildb_update_email(const char * userid, int userid_len, const char * email,
...
+    if (r != SQLITE_OK)
+        goto end;
+
+    while (sqlite3_step(Stmt) == SQLITE_ROW) {
+        userec_t u = {};
+        const char *userid = (const char *)sqlite3_column_text(Stmt, 0);
+        int64_t generation = sqlite3_column_int64(Stmt, 1);
+
+        if (userid == NULL)
+            continue;  // Bad record.
```
Comment (wens) [inline daemon/regmaild/regmaild.c:413]: You can get rid of this case by checking for `userid IS NOT NULL` in the `WHERE` clause.

File: daemon/regmaild/regmaild.c [hunk 18]
```diff
@@ -607,6 +849,47 @@ client_cb(int fd, short event, void *arg)
...
+        }
+
+        case VERIFYDB_REQ_GET:
+        {
+            const verifydb_req *req = &storage.verifydb;
+            if (!validate_verifydb_request(req))
+                goto end;
+
+            verifydb_get(req, fd);
+            goto end_noclose;
```
Comment (wens) [inline daemon/regmaild/regmaild.c]: Why the difference? Please keep all handling consistent.

**Comment by wens** [comment]
For `regmailc: Add client support for verifydb.` please also split up the cleanup/restructure and verifydb additions.

**Comment by wens** [comment]
Reviewed-by for the following:
- regmaild: Use "\n" as new line, instead of "\r\n".
- regmaild: Error and exit if DB cannot be opened.
- regmaild: Use g_Db in handler.
- regmaild: Factor out request validation into utility function.
- regmaild: Prepare for accepting different request struct per operation.
- regmaild: Return non-zero exit code on init error.
- regmaild: Refine error messages; nit style fixes.
- regmaild: Add initemaildb into list of programs.

Please push these out to master and rebase, so we don't have to look at them again.

For `Fix include in daemons.h.`, it looks like you squashed in something wrong.

File: mbbsd/register.c [hunk 19]
```diff
@@ -1264,7 +1262,7 @@ create_regform_request()
 }
 
 static int
-register_email_input(const char *userid, char *email)
+register_email_input(const userec_t *u, char *email)
```
Comment (wens) [inline mbbsd/register.c:1293]: Within the commit `register: Move userid hack into emaildb functions.`, the parameter type change is unwarranted.
Please move this to the commit that actually needs it, which I assume is `register: Dual read/write to verifydb.`

**Comment by wens** [comment]
Also, you don't need a full stop (period) at the end of the commit subject.
It is a subject line, much like for an article, not a full sentence, so the full stop is not needed.

File: daemon/regmaild/regmaild.c (hunk 16 above)
Comment (robertabcd) [reply daemon/regmaild/regmaild.c]: My bad. It should return the record though. I made it return an empty string.
Right now, we won't have this case.

Comment (robertabcd) [reply daemon/regmaild/regmaild.c]: There may be multiple records. It needs the count in the reply struct that is going to be prepended.

File: daemon/regmaild/regmaild.c (hunk 18 above)
Comment (robertabcd) [reply daemon/regmaild/regmaild.c]: Fixed.

File: mbbsd/register.c (hunk 19 above)
Comment (robertabcd) [reply mbbsd/register.c:1293]: Moved.

**Comment by robertabcd** [comment]
> Also, you don't need a full stop (period) at the end of the commit subject.
> It is a subject line, much like for an article, not a full sentence, so the full stop is not needed.

Fixed.

File: include/daemons.h (hunk 15 above)
Comment (robertabcd) [reply include/daemons.h:117]:
```c
typedef struct {
  regmaildb_req_header;
  int xxx;
} test_t;

void f() {
    test_t t;
    t.cb = 100;
}
```

```
../../include/daemons.h:120:3: warning: declaration does not declare anything [-Wmissing-declarations]
  regmaildb_req_header;
  ^~~~~~~~~~~~~~~~~~~~

regmaild.c:890:7: error: no member named 'cb' in 'test_t'
    t.cb = 100;
    ~ ^
```

Comment (wens) [reply include/daemons.h:117]: Ah, ok, so anonymous members don't work this way. Too bad.

**Comment by robertabcd** [comment]
Discarding this. Working on a flatbuffers protocol.
